File: personas/generator.py
"""Generate .claude.md files for agent personas."""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from plan.models import ProjectPlan, Task, Persona
from .models import PersonaConfig, AgentInstance


class ClaudeMdGenerator:
    """Generates .claude.md files for agent sandboxes."""

    DEFAULT_TEMPLATE = """# {name} Agent

## My Role
I am a **{role}** agent in a multi-agent orchestration system.

## My Mission
{mission}

## My Capabilities
{capabilities}

## My Constraints
{constraints}

## Current Task
{current_task}

## Communication Protocol

I communicate with other agents via:
1. **Git commits** - For code delivery
2. **Pull requests** - For handoff to integration branch
3. **communications.json** - For status updates and requests

### Signaling Breakpoints

When I reach a natural stopping point, I update communications.json:

**Task Complete:**
```json
{{
    "lifecycle_state": "complete",
    "done": "Description of what I completed",
    "breakpoint": {{
        "type": "task_complete",
        "task_id": "T001",
        "summary": "Brief summary of work done"
    }}
}}
```

**Blocked:**
```json
{{
    "lifecycle_state": "blocked",
    "working_on": "What I'm waiting for",
    "breakpoint": {{
        "type": "blocked",
        "blocked_on": ["T002", "T003"],
        "reason": "Need X before I can continue"
    }}
}}
```

**PR Created:**
```json
{{
    "lifecycle_state": "pr_pending",
    "breakpoint": {{
        "type": "pr_created",
        "pr_url": "https://github.com/..."
    }}
}}
```

## Testing Philosophy

- Write **E2E tests only** - no unit tests
- Test through the **public API** only
- Each acceptance criterion must have a passing test
- Target **100% coverage** on all metrics
- If code can't be reached by a real user scenario, it's dead code

## Git Workflow

1. Work on branch: `{branch}`
2. Make small, atomic commits
3. When task is complete, create PR to integration branch
4. Signal breakpoint in communications.json
5. Wait for orchestrator to handle next steps

## Remember

- I work ONLY in the sandbox directory
- I follow the existing code style
- I don't over-engineer - just what's needed
- I signal breakpoints at natural stopping points
- Fresh context will be provided if I'm respawned
"""

    def __init__(self, plan: ProjectPlan):
        self.plan = plan

    def generate(
        self,
        persona_config: PersonaConfig,
        task: Optional[Task] = None,
        branch: str = "",
        context_summary: str = "",
    ) -> str:
        """
        Generate .claude.md content for an agent.

        Args:
            persona_config: The persona configuration
            task: The current task (if any)
            branch: The git branch for this agent
            context_summary: Additional context from previous runs
        """
        persona = persona_config.persona

        # Use custom template if available, otherwise default
        template = persona.claude_md_template or self.DEFAULT_TEMPLATE

        # Format capabilities
        capabilities = "\n".join(f"- {cap}" for cap in persona.capabilities) or "- General development"

        # Format constraints
        constraints = "\n".join(f"- {con}" for con in persona.constraints) or "- Follow project conventions"

        # Format current task
        if task:
            current_task = f"""
**Task ID:** {task.id}
**Description:** {task.description}
**Dependencies:** {', '.join(task.dependencies) if task.dependencies else 'None'}
"""
        else:
            current_task = "No task currently assigned. Check communications.json for available tasks."

        # Build mission from persona and plan context
        mission = self._build_mission(persona_config)

        content = template.format(
            name=persona.name,
            role=persona.role,
            mission=mission,
            capabilities=capabilities,
            constraints=constraints,
            current_task=current_task,
            branch=branch or "agent/unknown",
        )

        # Add context summary if provided (for respawned agents)
        if context_summary:
            content += f"""
## Context from Previous Run

{context_summary}
"""

        return content

    def _build_mission(self, persona_config: PersonaConfig) -> str:
        """Build a mission statement for the persona."""
        persona = persona_config.persona
        role = persona.role

        # Role-specific mission templates
        missions = {
            "architect": "Design system architecture and make technical decisions that guide the project.",
            "implementer": "Implement features based on specifications, following best practices.",
            "tester": "Write comprehensive E2E tests that verify acceptance criteria.",
            "documenter": "Create clear, accurate documentation for the codebase.",
            "reviewer": "Review code for quality, security, and adherence to standards.",
        }

        return missions.get(role, f"Contribute to the project as a {role}.")

    def write_to_sandbox(
        self,
        sandbox_path: Path,
        persona_config: PersonaConfig,
        task: Optional[Task] = None,
        branch: str = "",
        context_summary: str = "",
    ):
        """
        Write .claude.md to a sandbox directory.

        Args:
            sandbox_path: Path to the sandbox directory
            persona_config: The persona configuration
            task: The current task (if any)
            branch: The git branch for this agent
            context_summary: Additional context from previous runs
        """
        sandbox_path = Path(sandbox_path)
        sandbox_path.mkdir(parents=True, exist_ok=True)

        content = self.generate(persona_config, task, branch, context_summary)

        claude_md_path = sandbox_path / ".claude.md"
        claude_md_path.write_text(content)

        return claude_md_path

    def generate_context_summary(self, agent: AgentInstance) -> str:
        """
        Generate a context summary for a respawned agent.
        This captures the essential state from the previous run.
        """
        lines = []

        lines.append(f"**Spawn #{agent.spawn_count}** - {datetime.now().isoformat()}")
        lines.append("")

        # Previous task progress
        if agent.current_task:
            lines.append("### Previous Task")
            lines.append(f"- **Task:** {agent.current_task.id} - {agent.current_task.description}")
            lines.append(f"- **Status:** {agent.current_task.status.value}")
            lines.append("")

        # Breakpoint info
        if agent.breakpoint:
            lines.append("### Last Breakpoint")
            lines.append(f"- **Type:** {agent.breakpoint.type}")
            if agent.breakpoint.summary:
                lines.append(f"- **Summary:** {agent.breakpoint.summary}")
            if agent.breakpoint.blocked_on:
                lines.append(f"- **Blocked on:** {', '.join(agent.breakpoint.blocked_on)}")
            if agent.breakpoint.reason:
                lines.append(f"- **Reason:** {agent.breakpoint.reason}")
            lines.append("")

        # Git state
        if agent.commits:
            lines.append("### Git History")
            for commit in agent.commits[-5:]:  # Last 5 commits
                lines.append(f"- {commit}")
            lines.append("")

        # Retry info
        if agent.retry_count > 0:
            lines.append(f"### Retry Info")
            lines.append(f"- **Previous retries:** {agent.retry_count}")
            lines.append("")

        return "\n".join(lines)

File: src/personas/generator.js
/**
 * @file Generator for .claude.md agent context files.
 * @module personas/generator
 */

/**
 * Generator for .claude.md files that provide context to Claude agents.
 */
export class ClaudeMdGenerator {
  /**
   * Create a ClaudeMdGenerator.
   * @param {import('../plan/models.js').ProjectPlan} plan - Project plan
   */
  constructor(plan) {
    /** @type {import('../plan/models.js').ProjectPlan} */
    this.plan = plan;
  }

  /**
   * Generate the .claude.md content for an agent.
   *
   * @param {import('./models.js').PersonaConfig} personaConfig - Agent's persona configuration
   * @param {import('../plan/models.js').Task} task - Current task
   * @param {string} branch - Git branch for this agent
   * @param {string} [contextSummary=''] - Summary from previous iterations
   * @returns {string} Generated .claude.md content
   */
  generate(personaConfig, task, branch, contextSummary = '') {
    const sections = [];

    // Header
    sections.push(this._generateHeader(personaConfig, task));

    // Mission
    sections.push(this._generateMission(personaConfig, task));

    // Context summary from previous iterations
    if (contextSummary) {
      sections.push(this._generateContextSection(contextSummary));
    }

    // Task details
    sections.push(this._generateTaskSection(task));

    // Capabilities and constraints
    sections.push(this._generateCapabilitiesSection(personaConfig));

    // Communication protocol
    sections.push(this._generateCommunicationSection(personaConfig.role, branch));

    // Breakpoint signaling
    sections.push(this._generateBreakpointSection(task));

    // Custom template content
    if (personaConfig.claudeMdTemplate) {
      sections.push(this._generateCustomSection(personaConfig.claudeMdTemplate));
    }

    return sections.join('\n\n');
  }

  /**
   * Generate header section.
   * @private
   * @param {import('./models.js').PersonaConfig} personaConfig
   * @param {import('../plan/models.js').Task} task
   * @returns {string}
   */
  _generateHeader(personaConfig, task) {
    return `# Agent: ${personaConfig.name}

**Role:** ${personaConfig.role}
**Current Task:** ${task.id}
**Project:** ${this.plan.name}`;
  }

  /**
   * Generate mission section.
   * @private
   * @param {import('./models.js').PersonaConfig} personaConfig
   * @param {import('../plan/models.js').Task} task
   * @returns {string}
   */
  _generateMission(personaConfig, task) {
    return `## Mission

You are "${personaConfig.name}", a ${personaConfig.role} agent working on the "${this.plan.name}" project.

Your current mission is to complete task **${task.id}**:
> ${task.description}`;
  }

  /**
   * Generate context summary section.
   * @private
   * @param {string} contextSummary
   * @returns {string}
   */
  _generateContextSection(contextSummary) {
    return `## Previous Progress

This is a continuation of previous work. Here's what has been accomplished:

${contextSummary}

**Important:** Build on this progress. Do not repeat completed work.`;
  }

  /**
   * Generate task details section.
   * @private
   * @param {import('../plan/models.js').Task} task
   * @returns {string}
   */
  _generateTaskSection(task) {
    let content = `## Task Details

**ID:** ${task.id}
**Description:** ${task.description}
**Role:** ${task.role}
**Status:** ${task.status}`;

    if (task.dependencies.length > 0) {
      content += `\n**Dependencies:** ${task.dependencies.join(', ')}`;
    }

    if (task.branch) {
      content += `\n**Branch:** ${task.branch}`;
    }

    // Find the story this task belongs to
    const story = this._findStoryForTask(task.id);
    if (story) {
      content += `\n\n### User Story

${story.userStoryText}`;

      if (story.acceptanceCriteria.length > 0) {
        content += `\n\n### Acceptance Criteria\n`;
        for (const ac of story.acceptanceCriteria) {
          content += `\n- **${ac.id}:** ${ac.description}`;
        }
      }
    }

    return content;
  }

  /**
   * Find the story containing a task.
   * @private
   * @param {string} taskId
   * @returns {import('../plan/models.js').Story|null}
   */
  _findStoryForTask(taskId) {
    for (const epic of this.plan.epics) {
      for (const story of epic.stories) {
        for (const task of story.tasks) {
          if (task.id === taskId) {
            return story;
          }
        }
      }
    }
    return null;
  }

  /**
   * Generate capabilities section.
   * @private
   * @param {import('./models.js').PersonaConfig} personaConfig
   * @returns {string}
   */
  _generateCapabilitiesSection(personaConfig) {
    let content = `## Your Capabilities`;

    if (personaConfig.capabilities.length > 0) {
      content += `\n\nYou are capable of:\n`;
      for (const cap of personaConfig.capabilities) {
        content += `- ${cap}\n`;
      }
    }

    if (personaConfig.constraints.length > 0) {
      content += `\n### Constraints\n\nYou must adhere to these constraints:\n`;
      for (const constraint of personaConfig.constraints) {
        content += `- ${constraint}\n`;
      }
    }

    return content;
  }

  /**
   * Generate communication protocol section.
   * @private
   * @param {string} role
   * @param {string} branch
   * @returns {string}
   */
  _generateCommunicationSection(role, branch) {
    return `## Communication Protocol

You coordinate with other agents via \`communications.json\`.

### Your Agent Entry

Update your status in communications.json under your agent ID. Include:
- **mission**: Your overall goal
- **workingOn**: Current activity
- **done**: Completed items
- **next**: Planned next steps

### Requesting Help

To request work from another agent, add to your \`requests\` array:
\`\`\`json
["target_agent_role", "description of what you need"]
\`\`\`

### Receiving Deliveries

Check your \`added\` array for completed work from other agents:
\`\`\`json
["from_agent", "what was delivered", "original request"]
\`\`\`

After processing deliveries, clear the \`added\` array.

### Git Branch

Work on branch: \`${branch}\`

Commit frequently with clear messages. Create a PR when the task is complete.`;
  }

  /**
   * Generate breakpoint signaling section.
   * @private
   * @param {import('../plan/models.js').Task} task
   * @returns {string}
   */
  _generateBreakpointSection(task) {
    return `## Breakpoint Signaling

When you reach a natural stopping point, signal a breakpoint by updating your status in communications.json:

### Task Complete
\`\`\`json
{
  "lifecycleState": "complete",
  "breakpoint": {
    "type": "task_complete",
    "taskId": "${task.id}",
    "summary": "Brief summary of what was accomplished"
  }
}
\`\`\`

### Blocked
\`\`\`json
{
  "lifecycleState": "blocked",
  "breakpoint": {
    "type": "blocked",
    "blockedOn": ["T001", "T002"],
    "reason": "Why you are blocked"
  }
}
\`\`\`

### PR Created
\`\`\`json
{
  "lifecycleState": "pr_pending",
  "breakpoint": {
    "type": "pr_created",
    "taskId": "${task.id}",
    "prUrl": "https://github.com/..."
  }
}
\`\`\`

**Important:** Always signal a breakpoint before stopping work. This allows the orchestrator to manage your lifecycle correctly.`;
  }

  /**
   * Generate custom template section.
   * @private
   * @param {string} template
   * @returns {string}
   */
  _generateCustomSection(template) {
    return `## Additional Instructions

${template}`;
  }

  /**
   * Generate a minimal .claude.md for quick tasks.
   *
   * @param {string} agentId - Agent identifier
   * @param {string} role - Agent role
   * @param {string} taskDescription - Task description
   * @param {string} branch - Git branch
   * @returns {string}
   */
  generateMinimal(agentId, role, taskDescription, branch) {
    return `# Agent: ${agentId}

**Role:** ${role}
**Branch:** ${branch}

## Task

${taskDescription}

## Communication

Update your status in \`communications.json\` under \`${agentId}\`.

Signal breakpoints when:
- Task complete: \`{"lifecycleState": "complete", "breakpoint": {"type": "task_complete", "summary": "..."}}\`
- Blocked: \`{"lifecycleState": "blocked", "breakpoint": {"type": "blocked", "blockedOn": [...], "reason": "..."}}\`
- PR created: \`{"lifecycleState": "pr_pending", "breakpoint": {"type": "pr_created", "prUrl": "..."}}\`
`;
  }
}
